<launch>
  <arg name="method" default="twist" />

  <!-- METHOD A: Standard Twist Filter (No Delay) -->
  <group if="$(eval arg('method') == 'twist')">
    <node pkg="fetch_teleop" type="safety_controller_twist.py" name="safety_twist" output="screen" />
    <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop_keyboard" output="screen">
      <remap from="cmd_vel" to="cmd_vel_teleop" />
    </node>
  </group>

  <!-- METHOD B: Runstop Trigger -->
  <group if="$(eval arg('method') == 'runstop')">
    <node pkg="fetch_teleop" type="safety_controller_runstop.py" name="safety_runstop" output="screen" />
    <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop_keyboard" output="screen">
    </node>
  </group>

  <!-- METHOD C: FULL LATENCY EXPERIMENT (Keyboard + Camera) -->
  <group if="$(eval arg('method') == 'delayed')">
    
    <!-- 1. The Safety Node (Listens to /cmd_vel_teleop) -->
    <node pkg="fetch_teleop" type="safety_controller_twist.py" name="safety_twist" output="screen" />

    <!-- 2. The Delay Node (Listens to /cmd_vel_raw and /head_camera/rgb/image_raw) -->
    <node pkg="fetch_teleop" type="cmd_delay.py" name="delay_node" output="screen" />

    <!-- 3. The Keyboard (Publishes to /cmd_vel_raw) -->
    <node pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py" name="teleop_keyboard" output="screen">
      <remap from="cmd_vel" to="cmd_vel_raw" />
    </node>

  </group>
</launch>